component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java

```diff
@@ -340,18 +347,35 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
             succeededAttempt = validateWithBackUpCodes(context, userToken, authenticatedUser);
         }
 
+        /*If the email address of the user is not saved in the user's profile below code will be extracting the validated
+        email address from the context and save it in user's profile. If saving failed it will save a property in the
+        message context*/
         if (succeededAttempt && isLocalUser) {
             String username = authenticatedUser.toFullQualifiedUsername();
+            String userEmail;
             try {
-                String userEmail = getEmailValueForUsername(username, context);
-                if (StringUtils.isBlank(userEmail)) {
-                    Object verifiedEmailObject = context.getProperty(EmailOTPAuthenticatorConstants.REQUESTED_USER_EMAIL);
-                    if (verifiedEmailObject != null) {
+                userEmail = getEmailValueForUsername(username, context);
+            } catch (EmailOTPException e) {
+                throw new AuthenticationFailedException("Failed to get the email claim when proceed the EmailOTP flow ", e);
+            }
+
+            if (StringUtils.isEmpty(userEmail)) {
```

Comment:
use isBlank

Reply:
let's add a debug log here explaning scenario and log any context info

Reply:
Fixed the issue and added logs